Write session cookie once

bigpresh opened the pull request:
Previously, every time we try to read from a session, `write_session_id` would generate a new session cookie to add to the response (overwriting the previously-generated one each time) - this is a bit silly and wasteful, but more importantly, causes `Too late to set another cookie, headers already built` errors if you try to access a session var from e.g. an `after` hook.

So, check if we have already added the appropriate session cookie to the response object, and if so, all is fine, do nothing.

bigpresh committed:
Add Dancer::Response->get_cookie

A convenience method to fetch a cookie we've set in the response object by name.

bigpresh committed:
If we already set session cookie, don't do it again

This will be a minor performance improvement, but more importantly, stops us
from generating "Too late to set another cookie, headers already built" errors
if you e.g. try to read from the existing session in an after hook.

bigpresh committed:
Tests that we can read session in an after hook

bigpresh committed:
Expect the new routes I added to test app

bigpresh committed:
Use Dancer::Cookies->cookie, not grubbing in the response

bigpresh committed:
Revert "Add Dancer::Response->get_cookie"

This reverts commit e2eabb77f6a5c9f4220ccbfe63c99fbe25b80d80.

skington commented on t/17_apps/02_load_app.t:
```diff
@@ -29,8 +29,8 @@ ok defined($forum), "app 'Forum' is defined";
-is @{ $test_app->registry->routes->{'get'} }, 16, 
-    "16 routes are defined in test app";
+is @{ $test_app->registry->routes->{'get'} }, 18, 
+    "18 routes are defined in test app";
```
Trivial, but Test::More::is() will tell you that it should have been 18 but isn't, so you don't need to mention that in the test title as well.

skington commented on t/08_session/03_http_requests.t:
```diff
@@ -55,6 +55,22 @@ Test::TCP::test_tcp(
             "session looks good for client $client";
 
+            $res = $ua->get("http://127.0.0.1:$port/session/after_hook/read");
+            ok($res->{success}, "Reading a session var in after hook worked");
```
You didn't try to write a session variable in that hook. You're testing the hook where you just read the session.

skington commented on t/lib/TestApp.pm:
```diff
@@ -59,6 +59,33 @@ get '/read_session' => sub {
+        session after_hook => "value changed in hook";
+        $response->content("Read " . session('after_hook'));
+    }
+};
```
If you want to have two routes, one of which tries to alter the session in an after hook and one of which doesn't, why not define (a) standard code which matches two routes, (b) an after hook that modifies the response for both routes, and then (c) a further after hook that fires only for one of the routes?

skington replied:
More importantly: if you revert the change to Dancer::Session::Abstract, do the tests still pass? If they don't, I'm very confused, because it looks like all you're doing is messing with an in-memory structure and then immediately returning part of it, which should obviously succeed. Do you perhaps need another route that just returns the contents of the session?

bigpresh replied:
The issue was that reading from the session would as a side effect generate a new cookie and push it into the response's cookies (overwriting any that we'd already put there).  That's a little inefficient, but more importantly means that if you read from a session in an after hook (e.g. as my code at $work was doing to pull a value from the session for logging purposes) the attempt to push the cookie again would explode - even though there was no need to try to update it.

However, looking at this again, I should add an additional test that makes sure that the changes made to the session did persist in the session, rather than not explode but not be flushed either (my concern is that the session stuff may also use an `after` hook to flush the session contents to the appropriate storage, and if so, that could fire before our `after` hook, in which case any changes we make would be silently discarded, which would be Not Good.

bigpresh committed:
Test that session var set in hook persisted.

Just for paranoia, make sure that the change to the session made in the after
hook did indeed get persisted into the session storage.

Just in case the session gets flushed by an after hook which fires before the
after hook we used to do the modification - don't want to think that the change
gets silently discarded, which the previous tests I added wouldn't have noticed.

bigpresh committed:
Support testing using the DBI session engine.

Optional, since Dancer::Session::DBI isn't part of the core Dancer distribution,
and requires external database configuration, but make it possible to run these
tests against it too by setting env vars.

bigpresh committed:
Run these tests against the Simple session engine too.

May as well, because why not?  It helps test Dancer::Session::Simple, not that
it's complex or particularly used much, but doesn't hurt to increase test
coverage.

bigpresh committed:
Update expected number of routes

bigpresh committed:
A bit less repetition here

bigpresh commented:
Right, the fix in this PR definitely works for us, but the tests I added pass with or without the fix.

It turns out, after much, much digging (thanks @skington for the help, too) that the problem only manifests if you use `send_file` and then try to access the session in an after handler (e.g. in some logging code which wants to log what happened and the user ID from the session).

So, I should be able to add a test which uses `send_file`, and see it fail without the fix and pass with the fix, and then I can be happy.

bigpresh committed:
Tests which actually fail without the fix.

The previous tests didn't fail without the fix added in 1a1c2346a1aee (and
tweaked in later commits), so they weren't actually very good tests!

After much, much debuggery, it turns out that we were only seeing the problem in
cases where send_file had been used.

The new test which uses send_file and an after hook that reads the session
afterwards does in fact die with the "Too late to set another cookie..." error
without the fix, and passes fine with the fix.

bigpresh committed:
Update expected number of TestApp routes.

bigpresh committed:
Accidentally left in before_serializer hook in test

Not needed.
